### PR TITLE
[feat] Support for running dependent tests with the asynchronous execution policy

### LINKED ISSUE #483: Support for tests that depend on each other
I'm interested in looking into whether ReFrame is a good tool for supporting the EasyBuild (http://easybuilders.github.io/easybuild/) regression test, in which we rebuild each easyconfig in the central repository prior to making a new release.

Since easyconfigs depend on each other, this requires being able to specify dependencies between tests, e.g. test C depends on test A & B having run (and succeeded).

As far I can tell, this is not supported yet by ReFrame right now?

cc @victorusu @gppezzi

### PR DESCRIPTION
**Pull request by ekouts**
Fixes #483.

### CODE REVIEW

**Comment by vkarak** [comment]
@jenkins-cscs retry daint

File: reframe/frontend/executors/policies.py [hunk 1]
```diff
@@ -224,19 +262,49 @@ def runcase(self, case):
...
+                self._retired_tasks.remove(task)
+
     def _poll_tasks(self):
         '''Update the counts of running checks per partition.'''
         getlogger().debug('updating counts for running test cases')
         getlogger().debug('polling %s task(s)' % len(self._running_tasks))
         for t in self._running_tasks:
             t.poll()
 
+    def _poll_waiting(self):
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: Don't call this `_poll_xxx`, because you are not actually polling the tasks. I'd call it `_notify_waiting()`, since it notifies all the tasks waiting on dependencies.

File: reframe/frontend/executors/policies.py [hunk 2]
```diff
@@ -224,19 +262,49 @@ def runcase(self, case):
...
+        for partname, _ in self._running_tasks_counts.items():
+            for task in self._waiting_tasks[partname]:
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: Why a double loop here? The following should be enough:

```python
for partname, task in self._waiting_tasks.items():
```

As suggested before, remove the tasks from the list in a separate list.

File: reframe/frontend/executors/policies.py [hunk 3]
```diff
@@ -264,11 +331,19 @@ def _failall(self, cause):
         except IndexError:
             pass
 
+        for waiting_list in self._waiting_tasks.values():
+            getlogger().debug("waiting list size: %s" % len(waiting_list))
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: Single quotes.

File: reframe/frontend/executors/policies.py [hunk 4]
```diff
@@ -264,11 +331,19 @@ def _failall(self, cause):
...
+                task.abort(cause)
+
         for ready_list in self._ready_tasks.values():
             getlogger().debug('ready list size: %s' % len(ready_list))
             for task in ready_list:
                 task.abort(cause)
 
+        for task in self._completed_tasks:
+            task.abort(cause)
+
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: Move that after aborting the `_retired_tasks`.

File: reframe/frontend/executors/policies.py [hunk 5]
```diff
@@ -307,7 +382,8 @@ def exit(self):
         pollrate = PollRateFunction(0.2, 60)
         num_polls = 0
         t_start = datetime.now()
-        while self._running_tasks or self._retired_tasks:
+        while (self._running_tasks or self._completed_tasks
+               or self._retired_tasks):
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: Put every condition in a separate line here.

File: reframe/frontend/executors/__init__.py [hunk 6]
```diff
@@ -145,13 +145,19 @@ def failed(self):
     def failed_stage(self):
         return self._failed_stage
 
+    @property
+    def succeeded(self):
+        return (self._current_stage == 'finalize'
+                or self._current_stage == 'cleanup')
```
Comment (vkarak) [inline reframe/frontend/executors/__init__.py]: Better write this as follows:

```python
return self._current_stage in {'finalize', 'cleanup'}
```

**Comment by vkarak** [comment]
We might also need a couple more unit tests to cover some corner cases.

File: reframe/frontend/executors/policies.py [hunk 7]
```diff
@@ -183,11 +200,34 @@ def runcase(self, case):
...
+        self._waiting_tasks.setdefault(partition.fullname, [])
         self._max_jobs.setdefault(partition.fullname, partition.max_jobs)
 
         task = RegressionTask(case, self.task_listeners)
+        self._task_index[case] = task
         self.stats.add_task(task)
+        self.printer.status(
+            'RUN', '%s on %s using %s' %
+            (check.name, partition.fullname, environ.name)
+        )
```
[the remark this answers is not in the document]
Comment (ekouts) [reply reframe/frontend/executors/policies.py]: Do we want to print the `RUN _ on _ using _` message when all dependencies of the testcase have already failed? Should I print a different message when it fails immediately? Also, the setup function is also called in _notify_waiting(), which can be called after the message `[----------] waiting for spawned checks to finish` and there will be no `RUN _ on _ using _` message for this testcases.

[the remark this answers is not in the document]
Comment (vkarak) [reply reframe/frontend/executors/policies.py]: I'd say, keep it as you have it now, but print `DEP` instead of `WAIT`.

File: reframe/frontend/executors/policies.py [hunk 8]
```diff
@@ -224,19 +275,50 @@ def runcase(self, case):
...
+
     def _poll_tasks(self):
         '''Update the counts of running checks per partition.'''
         getlogger().debug('updating counts for running test cases')
         getlogger().debug('polling %s task(s)' % len(self._running_tasks))
         for t in self._running_tasks:
             t.poll()
 
+    def _setup_all(self):
+        tasks_to_remove = []
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: This name is a bit confusing and the statement at the end that you remove them from the waiting list. I would write it as follows:

```python
still_waiting = []
for stuff:
    if some_stuff:
        ...
    elif other_stuff:
        ...
    else:
        still_waiting.append(task)

self._waiting_tasks[:] = still_waiting
```

File: reframe/frontend/executors/policies.py [hunk 9]
```diff
@@ -272,6 +356,9 @@ def _failall(self, cause):
         for task in self._retired_tasks:
             task.abort(cause)
 
+        for task in self._completed_tasks:
+            task.abort(cause)
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: You could combine the killing of the waiting, retired and completed tasks in a single loop:

```python
for task in itertools.chain(self._waiting_tasks, self._retired_tasks, self._completed_tasks):
    task.abort(cause)
```

File: reframe/frontend/executors/policies.py [hunk 10]
```diff
@@ -307,13 +394,15 @@ def exit(self):
         pollrate = PollRateFunction(0.2, 60)
         num_polls = 0
         t_start = datetime.now()
-        while self._running_tasks or self._retired_tasks:
+        while (self._running_tasks or self._waiting_tasks):
```
Comment (vkarak) [inline reframe/frontend/executors/policies.py]: You will miss tasks here with this condition. As soon as you schedule your last batch of tasks, you won't be able to finalize them, because you will be out of the loop. That's why I was checking also the `retired_tasks` list. You should check that all the intermediate lists have be flushed.

Comment (ekouts) [reply reframe/frontend/executors/policies.py]: A task is removed from `running_tasks` inside `on_task_exit` or inside `on_task_failure`. In the first case it is called by `_poll_tasks()` so in the next loop. In the second case the failed task will not have to be cleaned up, as well as the tasks the failed task was depending on. So which case would be caught if I was checking for `completed_tasks` or `ready_tasks`? Not all completed tasks should be cleaned up in the end.

Comment (vkarak) [reply reframe/frontend/executors/policies.py]: I was thinking mostly about the case of success rather that failure, but your explanation makes sense.

**Comment by ekouts** [comment]
> The PR is very good, but there is still an issue to be addressed and is not caught by the unit tests currently. In the following, I have cancelled the execution before entering the final busy loop and none of the already processed tests is marked as failed:
> 
> ```
> $ ./bin/reframe -c unittests/resources/checks_unlisted/deps_complex.py --exec-policy=async --prefix=$SCRATCH/rfm-stage/ -r
> ./bin/reframe: could not find a configuration entry for the current system; falling back to a generic system configuration; please check the online documentation on how to configure ReFrame for your system.
> Command line: ./bin/reframe -c unittests/resources/checks_unlisted/deps_complex.py --exec-policy=async --prefix=/scratch/snx3000tds/karakasv/rfm-stage/ -r
> Reframe version: 2.21-dev0
> Launched by user: karakasv
> Launched on host: dom101
> Reframe paths
> =============
>     Check prefix      :
>     Check search path : 'unittests/resources/checks_unlisted/deps_complex.py'
>     Stage dir prefix     : /scratch/snx3000tds/karakasv/rfm-stage/stage/
>     Output dir prefix    : /scratch/snx3000tds/karakasv/rfm-stage/output/
>     Perf. logging prefix : /scratch/snx3000tds/karakasv/rfm-stage/perflogs
> [==========] Running 8 check(s)
> [==========] Started on Mon Nov 11 14:38:10 2019
> 
> [----------] started processing T0 (T0)
> [ RUN      ] T0 on generic:login using builtin-gcc
> [----------] finished processing T0 (T0)
> 
> [----------] started processing T4 (T4)
> [ RUN      ] T4 on generic:login using builtin-gcc
> [      DEP ] T4 on generic:login using builtin-gcc
> [----------] finished processing T4 (T4)
> 
> [----------] started processing T5 (T5)
> [ RUN      ] T5 on generic:login using builtin-gcc
> [      DEP ] T5 on generic:login using builtin-gcc
> [----------] finished processing T5 (T5)
> 
> [----------] started processing T1 (T1)
> [ RUN      ] T1 on generic:login using builtin-gcc
> [      DEP ] T1 on generic:login using builtin-gcc
> [----------] finished processing T1 (T1)
> 
> [----------] started processing T6 (T6)
> ^C[  PASSED  ] Ran 8 test case(s) from 8 check(s) (0 failure(s))
> [==========] Finished on Mon Nov 11 14:38:10 2019
> ```
> 
> Your implementation of the async policy seems to work correctly when no test with dependencies are used:
> 
> ```
> $ ./bin/reframe -C config/cscs.py -c cscs-checks/apps/ -R -t production --exec-policy=async --prefix=$SCRATCH/rfm-stage/ -r
> Command line: ./bin/reframe -C config/cscs.py -c cscs-checks/apps/ -R -t production --exec-policy=async --prefix=/scratch/snx3000tds/karakasv/rfm-stage/ -r
> Reframe version: 2.21-dev0
> Launched by user: karakasv
> Launched on host: dom101
> Reframe paths
> =============
>     Check prefix      :
> (R) Check search path : 'cscs-checks/apps/'
>     Stage dir prefix     : /scratch/snx3000tds/karakasv/rfm-stage/stage/
>     Output dir prefix    : /scratch/snx3000tds/karakasv/rfm-stage/output/
>     Perf. logging prefix : /scratch/snx3000tds/karakasv/rfm-stage/perflogs
> [==========] Running 16 check(s)
> [==========] Started on Mon Nov 11 14:40:53 2019
> 
> [----------] started processing GromacsGPUCheck_small_prod (GROMACS GPU check)
> [ RUN      ] GromacsGPUCheck_small_prod on dom:gpu using PrgEnv-gnu
> [----------] finished processing GromacsGPUCheck_small_prod (GROMACS GPU check)
> 
> [----------] started processing GromacsCPUCheck_small_prod (GROMACS CPU check)
> [ RUN      ] GromacsCPUCheck_small_prod on dom:mc using PrgEnv-gnu
> [----------] finished processing GromacsCPUCheck_small_prod (GROMACS CPU check)
> 
> [----------] started processing Cp2kCpuCheck_small_prod (CP2K CPU check (version: small, prod))
> [ RUN      ] Cp2kCpuCheck_small_prod on dom:mc using PrgEnv-gnu
> [----------] finished processing Cp2kCpuCheck_small_prod (CP2K CPU check (version: small, prod))
> 
> [----------] started processing Cp2kGpuCheck_small_prod (CP2K GPU check (version: small, prod))
> [ RUN      ] Cp2kGpuCheck_small_prod on dom:gpu using PrgEnv-gnu
> ^C[     FAIL ] Cp2kGpuCheck_small_prod on dom:gpu using PrgEnv-gnu
> [     FAIL ] Cp2kCpuCheck_small_prod on dom:mc using PrgEnv-gnu
> [     FAIL ] GromacsCPUCheck_small_prod on dom:mc using PrgEnv-gnu
> [     FAIL ] GromacsGPUCheck_small_prod on dom:gpu using PrgEnv-gnu
> [  FAILED  ] Ran 17 test case(s) from 16 check(s) (4 failure(s))
> [==========] Finished on Mon Nov 11 14:41:04 2019
> 
> ==============================================================================
> SUMMARY OF FAILURES
> ------------------------------------------------------------------------------
> FAILURE INFO for GromacsGPUCheck_small_prod
>   * System partition: dom:gpu
>   * Environment: PrgEnv-gnu
>   * Stage directory: /scratch/snx3000tds/karakasv/rfm-stage/stage/dom/gpu/PrgEnv-gnu/GromacsGPUCheck_small_prod
>   * Job type: batch job (id=834178)
>   * Maintainers: ['VH']
>   * Failing phase: run
>   * Reason: aborted due to KeyboardInterrupt
> ------------------------------------------------------------------------------
> FAILURE INFO for GromacsCPUCheck_small_prod
>   * System partition: dom:mc
>   * Environment: PrgEnv-gnu
>   * Stage directory: /scratch/snx3000tds/karakasv/rfm-stage/stage/dom/mc/PrgEnv-gnu/GromacsCPUCheck_small_prod
>   * Job type: batch job (id=834179)
>   * Maintainers: ['VH']
>   * Failing phase: run
>   * Reason: aborted due to KeyboardInterrupt
> ------------------------------------------------------------------------------
> FAILURE INFO for Cp2kCpuCheck_small_prod
>   * System partition: dom:mc
>   * Environment: PrgEnv-gnu
>   * Stage directory: /scratch/snx3000tds/karakasv/rfm-stage/stage/dom/mc/PrgEnv-gnu/Cp2kCpuCheck_small_prod
>   * Job type: batch job (id=834180)
>   * Maintainers: ['LM']
>   * Failing phase: run
>   * Reason: aborted due to KeyboardInterrupt
> ------------------------------------------------------------------------------
> FAILURE INFO for Cp2kGpuCheck_small_prod
>   * System partition: dom:gpu
>   * Environment: PrgEnv-gnu
>   * Stage directory: /scratch/snx3000tds/karakasv/rfm-stage/stage/dom/gpu/PrgEnv-gnu/Cp2kGpuCheck_small_prod
>   * Job type: batch job (id=None)
>   * Maintainers: ['LM']
>   * Failing phase: run
>   * Reason: cancelled by user
> ------------------------------------------------------------------------------
> ```

I think the problem is not an issue of the asynchronous implementation or dependencies. If the ^C happens outside the try part of `runcase` or inside a different part `_runall` the exception won't be caught in the right place and _failall() from the policy won't be called.

**Comment by vkarak** [comment]
Thanks @ekouts, you're right. This is not related to this PR. Great work! 🎉 I'm gonna merge it.
